Support cross compiling

Un1q32 opened the pull request:
Closes #48

Handles signum.h by pre-generating the header for different operating systems (does it ever differ by architecture?)
Handles tryexec tests by assuming success when cross compiling.
Handles operating system detection by requiring the user to specify it with an environment variable.
Handles POSIX/XOPEN versions by guessing 200112 and 600.
Made the getcwd(NULL, 0) check default to not supported to be safe, and added an environment variable to override it.
Added the --host option to configure, this adds a cross prefix to cc and ar, so if you pass --host=aarch64-linux-musl then configure will look for aarch64-linux-musl-gcc and aarch64-linux-musl-ar

Un1q32 committed:
get rid of stupid macos debug files

Un1q32 committed:
cross compiling

Un1q32 committed:
support --host option

Un1q32 committed:
make getcwd(NULL,0) check default to false, and provide variable to control result

Un1q32 committed:
add freebsd signum

Un1q32 committed:
add openbsd signum

silvanshade commented:
Thanks for this.

I have a use case where I need cross-compilation for `yash` and this PR seems to work.

One suggested addition would be to support the usual [`--build`](https://gcc.gnu.org/onlinedocs/gccint/Configure-Terms.html) option for configure:

```
./configure --build=x86_64-unknown-linux-gnu --host=aarch64-unknown-linux-gnu
```

The `--build` option will often be passed by default when cross-compiling with various tooling, such as [nix](https://github.com/NixOS/nix), and the OS can be determined from this value instead of needing a separate environment var.

Un1q32 commented:
> the OS can be determined from this value instead of needing a separate environment var.

`--build` specifies the system we're building on, not the target.  The $OS environment variable specifies the target OS.
I don't think there's anywhere in the configure script that needs to know a triple for the build environment, but I can make the build option be accepted but ignored for autoconf compatibility.

Determining the OS from the triple would be complex, since the triple is in the format {cpu}-{vendor}-{os}{version}-{abi} but sometimes the abi or vendor or both are ommited, so for an example I don't know an easy way to tell if x86_64-linux-gnu means an OS of linux or gnu, or if arm64-apple-darwin22 means an os of darwin or apple.  I think autoconf solves this by having a list of known values used as the vendor to check if the os is the 2nd or 3rd field, but then the list would have to be continuously updated and it's a lot easier to just make the user specify it in an environment variable.

Un1q32 committed:
add --build stub option for autoconf compatability

silvanshade commented:
> > the OS can be determined from this value instead of needing a separate environment var.
> 
> `--build` specifies the system we're building on, not the target.

Right.

> The $OS environment variable specifies the target OS. I don't think there's anywhere in the configure script that needs to know a triple for the build environment, but I can make the build option be accepted but ignored for autoconf compatibility.

Okay, I see now.

When I first tried using this, I was cross-compiling with nix, which as I mentioned passes the `--build=` flag by default, which in this case gave an error. After modifying the build script to omit that flag, then the yash configure complained about not being able to determine the OS, and I just assumed (though it seemed odd) that it had something to do with `--build`.

Maybe it would be clearer if it were named `$HOST_OS` or even `$YASH_HOST_OS`?

> Determining the OS from the triple would be complex, since the triple is in the format {cpu}-{vendor}-{os}{version}-{abi} but sometimes the abi or vendor or both are ommited, so for an example I don't know an easy way to tell if x86_64-linux-gnu means an OS of linux or gnu, or if arm64-apple-darwin22 means an os of darwin or apple.

Theoretically, I agree that extracting the correct information for the triple with perfect accuracy has a number of issues.

In practice, I don't think it's really a problem though, due to the regularity and small set of actual triples that are encountered on a normal basis.

I think the vast majority will be of the form `{arch}-apple-darwin` or `{arch}-unknown-freebsd` or `{arch}-unknown-linux-{gnu,musl}`. I don't know that the `darwin` vs `apple` distinction is really meaningful for this case, is it?

A best-effort heuristic could catch probably 95% of the cases and then for the remaining ambiguous cases you could fall back to requiring the explicit environment variable or another flag or something.

But it's just a suggestion.

> I think autoconf solves this by having a list of known values used as the vendor to check if the os is the 2nd or 3rd field, but then the list would have to be continuously updated and it's a lot easier to just make the user specify it in an environment variable.

True, it's not a huge burden, but the downside is that it still requires an extra configuration step for consumers.

For the case of nix which I was referencing before, the yash cross-compile would have "just worked" (with this patch) like most other autoconf-based packages, if:

1. the `--build=` flag were accepted (even if ignored)
2. the `--host=` flag were used to determine the host OS

Un1q32 commented:
> I think the vast majority will be of the form `{arch}-apple-darwin` or `{arch}-unknown-freebsd` or `{arch}-unknown-linux-{gnu,musl}`. I don't know that the `darwin` vs `apple` distinction is really meaningful for this case, is it?

Omitting vendor is extremely common, like half of all linux cross toolchains do it, and the remaining have lots of different vendors, unknown, pc, redhat, and suse are just 4 of the ones I know of.  The only solution is to have either a list of allowed operating systems or a list of allowed vendors that we can use to check which fields are which.  Failing only on edge cases isn't viable either since we would still need to have a list of operating systems or vendors to determine what is and isn't an edge case.

silvanshade commented:
> Omitting vendor is extremely common, like half of all linux cross toolchains do it, and the remaining have lots of different vendors, unknown, pc, redhat, and suse are just 4 of the ones I know of.

I guess I don't understand why this is a problem.

Why does whether or not the vendor is included in the triple matter for determining the OS? Are you intending to eventually use that information for some implementations?

> Failing only on edge cases isn't viable either since we would still need to have a list of operating systems or vendors to determine what is and isn't an edge case.

Isn't that already effectively the case given that there OS-specific signum implementations in the PR?

Consider a very simple approach like this:

    if `OS` is set, use that value
    else if `--host=<triple>` contains substring `darwin` then `OS=darwin`
    else if `--host=<triple>` contains substring `freebsd` then `OS=freebsd`
    else if `--host=<triple>` contains substring `linux` then `OS=linux`
    else if `--host=<triple>` contains substring `openbsd` then `OS=openbsd`
    otherwise error and request user set `OS`

Wouldn't that work?

Given that (as far as I'm aware) no triple contains multiple OS values, that should always be unambiguous, right?

Un1q32 commented:
I don't want to do that because as it is right now all you need to do to allow a new OS to be cross compiled for is put a file in the root of the build folder called `<os>-signum.h`, but having a list would mean you also have to update that list and that's more work.  Maybe I could auto generate the list of allowed OSes by checking what *-signum.h files exist.

silvanshade commented:
> but having a list would mean you also have to update that list and that's more work.

Fair point.

> Maybe I could auto generate the list of allowed OSes by checking what *-signum.h files exist.

This seems like it should suffice, and cheap enough to generate on the fly in the configure script just from the file listings and then checking for the `*-signum.h` prefixes as substrings in `host`.

silvanshade committed:
determine target OS from --host

silvanshade commented:
> > but having a list would mean you also have to update that list and that's more work.
> 
> Fair point.
> 
> > Maybe I could auto generate the list of allowed OSes by checking what *-signum.h files exist.
> 
> This seems like it should suffice, and cheap enough to generate on the fly in the configure script just from the file listings and then checking for the `*-signum.h` prefixes as substrings in `host`.

I added a [commit](https://github.com/magicant/yash/commit/fcbff8ba6bbbcd0cc6570541ae82554624723d44) that implements this in my local branch: [silvanshade/yash:target-os-from-host](https://github.com/silvanshade/yash/tree/target-os-from-host)

Un1q32 commented:
I could probably do it without using find or grep, I'll try later.

magicant commented on configure:
```diff
@@ -341,10 +350,83 @@ cscope="${CSCOPE-cscope}"
+    ostype="$OS"
+fi
+
+if [ -n "$cross_compiling" ]; then
```
Can you assign `true` or `false` to `cross_compiling` so we can write `if ${cross_compiling}` instead of `if [ -n "$cross_compiling" ]`?

magicant commented on configure:
```diff
@@ -1439,19 +1488,22 @@ char *xgetcwd(void) {
 if [ x"${checkresult}" = x"yes" ]
 then
-    defconfigh "GETCWD_AUTO_MALLOC"
+    defconfigh "GETCWD_NO_AUTO_MALLOC"
```
> Made the getcwd(NULL, 0) check default to not supported to be safe, and added an environment variable to override it.

Yes, we should not assume that `getcwd(NULL, 0)` uses `malloc` when cross-compiling. However, could you revert to the existing macro name `GETCWD_AUTO_MALLOC`? Macros in config.h are defined when features are available or enabled (except for `*_BROKEN`), and this one should align with that convention.

magicant reviewed (commented):
Thank you so much for the pull request. I probably wouldn't have implemented this myself.

Un1q32 committed:
make $cross_compiling true/false

Un1q32 committed:
change back some stuff with the getcwd check

Un1q32 committed:
nit

magicant commented on path.c:
```diff
@@ -248,6 +245,9 @@ char *xgetcwd(void)
     return pwd;
+#else
+    char *pwd = getcwd(NULL, 0);
+    return (pwd != NULL) ? xrealloc(pwd, add(strlen(pwd), 1)) : NULL;
```
Please also update these lines to match the macro name `GETCWD_AUTO_MALLOC`.

magicant commented on configure:
```diff
@@ -1414,7 +1464,8 @@ fi
 # check if getcwd accepts (NULL,0) as argument
 checking "if getcwd(NULL,0) returns malloc'ed string"
-cat >"${tempsrc}" <<END
+if [ -z "$YASH_GETCWD_AUTO_MALLOC" ]; then
```
Can you add this variable to the list of "Influential environment variables" in the help message?

magicant commented on configure:
```suggestion
            if printf '%s\n' "$i" | grep -Eq "^[[:alpha:]][[:alnum:]_]*="
```

This should allow specifying the `YASH_GETCWD_AUTO_MALLOC` variable as an argument to the script as in `./configure YASH_GETCWD_AUTO_MALLOC=yes`.

Un1q32 committed:
fix

Un1q32 committed:
update env vars in --help

Un1q32 committed:
update line

silvanshade commented:
> > > > Maybe I could auto generate the list of allowed OSes by checking what *-signum.h files exist.
> > > 
> > > This seems like it should suffice, and cheap enough to generate on the fly in the configure script just from the file listings and then checking for the `*-signum.h` prefixes as substrings in `host`.
> > 
> > I added a [commit](https://github.com/magicant/yash/commit/fcbff8ba6bbbcd0cc6570541ae82554624723d44) that implements this in my local branch: [silvanshade/yash:target-os-from-host](https://github.com/silvanshade/yash/tree/target-os-from-host)

> I could probably do it without using find or grep, I'll try later.

@Un1q32 Are you still intending to add this?

The configure script already uses `find` and `grep` so it seems like it should be fine to use them here.

Un1q32 committed:
Merge remote-tracking branch 'other/target-os-from-host' into trunk

Un1q32 committed:
style nits

Un1q32 commented:
> @Un1q32 Are you still intending to add this?
> 
> The configure script already uses `find` and `grep` so it seems like it should be fine to use them here.

Never got around to it, probably never will, I merged your solution in for now at least.

magicant commented on configure:
```diff
@@ -341,10 +352,89 @@ cscope="${CSCOPE-cscope}"
-ostype=$(uname -s | tr "[:upper:]" "[:lower:]")
+if [ -n "$OS" ]; then
+    ostype="$OS"
+elif [ -n "$cross_compiling" ]; then
```
Since 86121d3a4c4d7f2c3e680b89e5091962faf2ec39, `$cross_compiling` expands to `true` or `false`.

```suggestion
elif $cross_compiling; then
```

magicant commented on configure:
```diff
@@ -341,10 +352,89 @@ cscope="${CSCOPE-cscope}"
+    ostype="$OS"
+elif [ -n "$cross_compiling" ]; then
+    for signum in $(find . -maxdepth 1 -name '*-signum.h' | sed 's/\.\/\(.*\)-signum\.h/\1/')
+    do
```
`-maxdepth` is not POSIX, so please avoid using it.

```suggestion
    for signum in *-signum.h
    do
        signum="${signum%-signum.h}"
```

magicant commented on configure:
```diff
@@ -341,10 +352,89 @@ cscope="${CSCOPE-cscope}"
+        then
+            ostype="$signum"
+            break;
+        fi
```
I guess your intention is to match `$signum` against `$host` literally. If so, you should use `grep -F` instead of `grep -E`.

There is another approach using `case`, which does not even use `grep`:

```suggestion
        case "$host" in
            *"$signum"*)
                ostype="$signum"
                break
        esac
```

Un1q32 committed:
Update configure

Co-authored-by: WATANABE Yuki <magicant@wonderwand.net>

Un1q32 committed:
Update configure

Co-authored-by: WATANABE Yuki <magicant@wonderwand.net>

Un1q32 committed:
Update configure

Co-authored-by: WATANABE Yuki <magicant@wonderwand.net>

Un1q32 committed:
fix case where $OS was defined when cross compiling, signum wouldn't be copied

Un1q32 committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

Un1q32 commented:
Gotta update the diagnostic messages a bit before merge, make it clear that if the os can't be determined it might be because of an absence of a signum.h, you currently have to dig through the configure sources to find that if you wanna port to a new os. I'll do it when I get home.

Un1q32 commented:
I forgot :sob:

Un1q32 committed:
improve error messages

Un1q32 committed:
better wording

Un1q32 commented:
The errors are a little longer than I'd like but I can't think of any way to shorten them without leaving out important info, open to suggestions.

magicant reviewed (approved):
LGTM

Un1q32 commented:
I've discovered signum.h DOES differ by architecture at least on linux.
This is the mips32 linux signum.h, note the larger RTSIZE than x86_64.
I don't believe this is true of other OSes like freebsd and darwin.
```c
/* signum.h: generated by makesignum */

#ifndef SIGNUM_H
#define SIGNUM_H

#include <stddef.h>

/* an injective function that returns an array index
 * corresponding to the given signal number,
 * which must be a valid non-realtime signal number
 * or zero. */
__attribute__((const))
static inline size_t sigindex(int signum) {
    return (size_t) signum;
}

/* max index returned by sigindex + 1 */
#define MAXSIGIDX 32

/* number of realtime signals that can be handled by yash */
#define RTSIZE 102

#endif
```

Un1q32 committed:
make linux signums arch dependent

Un1q32 commented:
There are 3 unique signums that I found, all differing only in RTSIZE, x86_64 has 40, mips has 102, and everything else has 39.

magicant commented:
Thanks, but now I'm thinking if we can eliminate automatic generation of signum.h at the cost of runtime efficiency. To be honest, I'd rather avoid maintaining all these platform-specific details.

Un1q32 committed:
remove changes to signum stuff

Un1q32 committed:
i dont think the test is needed

magicant committed:
Generic cross-signum.h

magicant commented:
I made a cross-signum.h that should work on all targets (but less efficiently) on my branch. Does it seem to work for you? https://github.com/magicant/yash/commit/cd9ee406909668d0013786bb1d4268973e1fe411

Un1q32 commented:
Any reason you can't do 4d33fa3f33d3bc163bb215a92499faf625d2490c?  Would improve efficiency of the cross build.

magicant commented:
> Any reason you can't do [4d33fa3](https://github.com/magicant/yash/commit/4d33fa3f33d3bc163bb215a92499faf625d2490c)? Would improve efficiency of the cross build.

SIGRTMAX and SIGRTMIN are not guaranteed to be compile-time constants. If those macros expand to something like `(__get_current_sigrtmax())`, they cannot be used as the size of `rttrap_command`, which is a statically allocated array.

Un1q32 commented:
There could be a configure check to see if the value would be constant when cross compiling, I might do that another time.  Other than that the commit looks fine except that makesignum should have a `2> /dev/null` to not show the error when it fails to run when cross compiling.

Un1q32 commented:
Wait couldn't you still do the
```c
#if defined SIGRTMIN && defined SIGRTMAX
// do whatever
#else
#define 0
#endif
```

Un1q32 commented:
nvm every use of RTSIZE is guarded by that condition anyway so it wouldn't matter.

Un1q32 committed:
redirect stderr

Un1q32 commented:
bug with cross compiling detection

Un1q32 committed:
fix cross compiling detection

the variable was unset when the first tryexec call happens which made it
always return success

Un1q32 committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

Un1q32 commented:
Unless anyone wants any more features I think this is complete.  I've tested cross compiling for mac from linux and everything works.  I can't think of anything more this needs.

Un1q32 commented:
With 0c40a88fa02d4d669ed48fd020254f920dba4142 it should be possible to make CI automatically test the cross build without needing to setup an emulator or anything, but I don't know enough about your CI setup to set that up myself.

Un1q32 commented:
Actually why is the default value for RTSIZE 100?  Does it work on MIPS Linux where the value generated by makesignum is 102?

magicant commented:
> Actually why is the default value for RTSIZE 100?  Does it work on MIPS Linux where the value generated by makesignum is 102?

Oops, just an oversight.

magicant commented on configure:
```diff
@@ -399,14 +399,20 @@ fi
-    tryexec
-then
-    checked "no"
+if [ -z "$cross_compiling" ]; then
```
If this is expected to be given as an environment variable, it should be all-caps and listed in the list of `Influential environment variables`.

magicant commented on configure:
```diff
@@ -399,14 +399,20 @@ fi
-    checked "yes"
     cross_compiling=true
+    checked "yes (forced)"
+    defconfigh "FORCE_CROSS_COMPILE"
```
> provide a way to make the build act like a cross build when its not for debugging
> 
> makes for an easy way to find issues with the cross built version
> without having to run an emulator or something

Please do not complicate things any more just to emulate something that could be done by manually editing Makefile.

magicant commented on makesignum.c:
```diff
@@ -79,14 +82,10 @@ int main(void)
-#else
-            0
+    wprintf(L"/* number of realtime signals that can be handled by yash */\n");
+    wprintf(L"#define RTSIZE %d\n\n", SIGRTMAX - SIGRTMIN + 10);
```
Could you revert these lines?
There's a chance that future me will forget that RTSIZE is defined conditionally, which would cause a subtle bug where compilation fails only in certain environments.

Un1q32 commented:
> > Actually why is the default value for RTSIZE 100?  Does it work on MIPS Linux where the value generated by makesignum is 102?
> 
> Oops, just an oversight.

What would a better default value be?

magicant commented:
> What would a better default value be?

I'd go with 255.

Un1q32 committed:
update

Un1q32 committed:
update list

Un1q32 committed:
a better approach to os detection

we dont actually *need* to always successfully detect an os for every
target, most oses don't affect the build later so we can just check for
every os that might affect the build and set the ostype to 'unknown' for
all others

Un1q32 commented:
Anything more you want from this?  I would consider this complete now.

magicant commented:
I'll merge this tomorrow.
